// bosch-smart-home-bridge/dist/model/bosch-client-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * This object contains the basic structure of a client during pairing
 *
 * @author Christopher Holomek
 * @since 26.09.2019
 */
var BoschClientData = /** @class */ (function () {
    /**
     * Create data for a new client with necessary information
     *
     * @param name
     *        name of the Open Source Project, which uses this library. "OSS " prefix is added automatically.
     * @param id
     *        identifier of the new client (pick a unique one). "oss_" prefix is added automatically.
     * @param certificate
     *        2048 bit selfsigned client certificate
     */
    function BoschClientData(name, id, certificate) {
        this.certificate = certificate;
        this['@type'] = 'client';
        this.primaryRole = 'ROLE_RESTRICTED_CLIENT';
        this.deleted = false;
        this.id = 'oss_' + id;
        this.name = 'OSS ' + name;
    }
    return BoschClientData;
}());
exports.BoschClientData = BoschClientData;
